src/main/antora/modules/security/pages/security.adoc

```diff
@@ -776,8 +776,352 @@ public class RestFormAuthCustomStoreIT extends ITBase {
+
+You'll learn how to:
+
+. Define security constraints
+. Define (and implicitly set) a custom authentication mechanism
+. Define (and implicitly set) a custom identity store
+. Use the Jakarta Security `SecurityContext`
+
+=== Write the application
+
+We'll use the same resource and same security constraints as we used for the <<Securing an endpoint with Basic authentication>> example.
+
+==== Define the authentication mechanism
+
+Let's now define a simple authentication mechanism that the security system can use to interact with the caller who tries to access a resource:
+
+[source,java]
+----
+@ApplicationScoped
+public class CustomAuthenticationMechanism implements HttpAuthenticationMechanism {
+    
+    @Inject
+    private IdentityStoreHandler identityStoreHandler;
+
+    @Override
+    public AuthenticationStatus validateRequest(
+        HttpServletRequest request, 
+        HttpServletResponse response, 
+        HttpMessageContext httpMessageContext) throws AuthenticationException {
+        
+        String callerName = request.getHeader("callername");
+        String password = request.getParameter("callerpassword");
+        
+        if (callerName == null || password == null) {
+            return httpMessageContext.doNothing();
+        }
+
+        CredentialValidationResult result = identityStoreHandler.validate(
+            new UsernamePasswordCredential(callerName, password));
+        
+        if (result.getStatus() != VALID) {
+            return httpMessageContext.responseUnauthorized();
+        }
+
+        return httpMessageContext.notifyContainerAboutLogin(
+            result.getCallerPrincipal(), 
+            result.getCallerGroups());
+    }
+    
+}
+----
+
+This custom authentication mechanism interacts with the caller by grabbing two headers from the request: `callername` and `callerpassword`. In case any of them are null, we return a special status; the "do nothing" status. This means there has been no request or attempt to do authentication. If the resource the caller is trying to access is not protected, the caller can access it anonymously. If it is proteced, the caller will not be able to access it.
+
+When the two required headers are provided by the caller, we create a `UsernamePasswordCredential` out of their values and pass that into the injected `IdentityStoreHandler`. We've seen how such handler worked in the example <<Securing an endpoint with Basic authentication and a custom algorithm for handling multiple identity stores>>.
+
+[NOTE]
+An authentication mechanism in Jakarta Security is not strictly required to delegate the credential validation to the identity store handler. However not doing so is considered bad practice, as it would restrict users from things like inserting extra identity stores into the chain that can do things like adding extra groups.
```

Comment:
When I read "users" here I think of the end user, not users of  the API.

Reply:
Good point, and that's why some of the EE docs introduced terms for them like "bean provider", "deployer", etc. Although these were kinda obscure to begin with.

I'll stick  with "developers" for now then.